### PR TITLE
refactor checkpoint loading for training type plugins

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7652

Motivation: Split out atomic responsibilities for restoring, and give training type plugins three hooks to customize:
- loading the checkpoint dict from a file
- setting model state dict
- setting optimizer state dict

so far these responsibilities were in one single hook and additionally duplicated the code in the CheckpointConnector for calling hooks on the LightningModule, DataModule, etc., such as `on_load_checkpoint`.

On top of that, the fact that it was entangled with a return value that determined whether to load optimizer state or not, would make it hard to do #7652 without this refactor. 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -133,7 +130,17 @@ def restore_model_state(self, model: LightningModule, checkpoint) -> None:
...
+        self.trainer.training_type_plugin.load_model_state_dict(checkpoint)
+
+    def restore_model_weights(self, checkpoint_path: Optional[Union[str, Path]]) -> None:
+        """ Restore only the model weights. """
+        checkpoint = self._loaded_checkpoint
+        if checkpoint_path is not None:
+            checkpoint = self.trainer.training_type_plugin.load_checkpoint_file(checkpoint_path)
+
+        model = self.trainer.lightning_module
+        model.on_load_checkpoint(checkpoint)
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:142]: nit:

```suggestion
        self.trainer.lightning_module.on_load_checkpoint(checkpoint)
```

**Review by carmocca — APPROVED** [review]
Love it!

**Review by SeanNaren — APPROVED** [review]
